p2p_beef_tx.go

```diff
@@ -0,0 +1,229 @@
+)
+
+const (
+	HasNoCMP = 0x00
+	HasCMP   = 0x01
+)
+
+type TxData struct {
+	Transaction *bt.Tx
+	PathIndex   *bt.VarInt
+}
+
+type DecodedBEEF struct {
+	CMPSlice        CMPSlice
+	InputsTxData    []TxData
+	ProcessedTxData TxData
+}
+
+func DecodeBEEF(beefHex string) (*DecodedBEEF, error) {
+	beefBytes, err := extractBytesWithoutVersionAndMarker(beefHex)
+	if err != nil {
+		return nil, err
+	}
+
+	cmpSlice, remainingBytes, err := DecodeCMPSliceFromStream(beefBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions, err := DecodeTransactionsWithPathIndexes(remainingBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(transactions) == 0 {
+		return nil, errors.New("no transactions found")
+	}
+
+	// get the last transaction as the processed transaction - it should be the last one because of khan's ordering
+	processedTx := transactions[len(transactions)-1]
+
+	transactions = transactions[:len(transactions)-1]
+
+	return &DecodedBEEF{
+		CMPSlice:        cmpSlice,
+		InputsTxData:    transactions,
+		ProcessedTxData: processedTx,
+	}, nil
+}
+
+func DecodeCMPSliceFromStream(hexBytes []byte) (CMPSlice, []byte, error) {
+	if len(hexBytes) == 0 {
+		return nil, nil, errors.New("provided hexStream is empty")
+	}
+
+	nCMPs, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
+	hexBytes = hexBytes[bytesUsed:]
+
+	var cmpPaths []CompoundMerklePath
+	for i := 0; i < int(nCMPs); i++ {
+		cmp, bytesUsedToDecodeCMP, err := NewCMPFromStream(hexBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cmpPaths = append(cmpPaths, cmp)
+		hexBytes = hexBytes[bytesUsedToDecodeCMP:]
+	}
+
+	cmpSlice := CMPSlice(cmpPaths)
+
+	return cmpSlice, hexBytes, nil
+}
+
+func NewCMPFromStream(hexBytes []byte) (CompoundMerklePath, int, error) {
+	height, bytesUsed, err := extractHeight(hexBytes)
+	if err != nil {
+		return nil, 0, err
+	}
+	hexBytes = hexBytes[bytesUsed:]
+
+	var cmp CompoundMerklePath
+	previousHeight := height
+	bytesUsedToDecodeCMP := bytesUsed
+
+	for previousHeight >= 0 {
+		var pathMap map[string]uint64
+
+		pathMap, bytesUsed, err = extractPathMap(hexBytes, previousHeight)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		cmp = append(cmp, pathMap)
+		hexBytes = hexBytes[bytesUsed:]
+
+		previousHeight--
+		bytesUsedToDecodeCMP += bytesUsed
+	}
+
+	return cmp, bytesUsedToDecodeCMP, nil
+}
+
+func DecodeTransactionsWithPathIndexes(bytes []byte) ([]TxData, error) {
+	nTransactions, offset := bt.NewVarIntFromBytes(bytes)
+	bytes = bytes[offset:]
+
+	var transactions []TxData
+
+	for i := 0; i < int(nTransactions); i++ {
+		tx, offset, err := bt.NewTxFromStream(bytes)
+		if err != nil {
+			return nil, err
+		}
+		bytes = bytes[offset:]
+
+		var pathIndex *bt.VarInt
+		if bytes[0] == HasCMP {
+			value, offset := bt.NewVarIntFromBytes(bytes[1:])
+			pathIndex = &value
+			bytes = bytes[1+offset:]
+		} else if bytes[0] == HasNoCMP {
+			bytes = bytes[1:]
+		} else {
+			return nil, errors.New("invalid HasCMP flag")
+		}
```

Comment:
If I understand correctly (which is not necessarily the case), ALL transactions in BEEF, other than the last one, need to have a CMP index. It is not mentioned anywhere in the BRC, but should we implement a check for this? Because right now we allow transactions (other than the last one that is being processed) that don't have CMP index.

Reply:
It's not true ;-) We use CMP to validate inputs from already mined transactions, but we can use inputs that come from transactions that are in the mempool only (not mined yet) as well.

We've decided to send only fully mined inputs for now in order to simplify and expedite the implementation and testing of this new mechanism.

In the real world, most BEEF transactions will consist of most or all of the transactions already mined, and so their verification will be simple and fast. We will have to do extended, and costly, verification only for the few inputs, not for all, which is a huge benefit